### PR TITLE
Element-R: implement encryption of outgoing events

### PR DESCRIPTION
**Pull request by richvdh**
This PR wires up the Rust-SDK into the event encryption path. I'm sorry it's a bit meaty - I've tried to break it down into bite-size commits and I recommend reviewing them in turn.

I'm also sorry that it's not as well tested as I would like, and in particular there are no integration tests. I've had trouble getting the tests to work for reasons including, but not limited to, https://github.com/matrix-org/matrix-rust-sdk/issues/1443. However, empirically it works in a browser, and I'm keen to see this shipped. I'll return to the tests soon.

---
Here's what your changelog entry will look like:

## ✨ Features
 * Element-R: implement encryption of outgoing events ([\#3122](https://github.com/matrix-org/matrix-js-sdk/pull/3122)).

### CODE REVIEW

File: src/rust-crypto/RoomEncryptor.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        await this.ensureEncryptionSession();
+
+        const encryptedContent = await this.olmMachine.encryptRoomEvent(
+            new RoomId(this.room.roomId),
+            event.getType(),
+            JSON.stringify(event.getContent()),
+        );
+
+        event.makeEncrypted(
+            "m.room.encrypted",
```
Comment (t3chguy) [inline src/rust-crypto/RoomEncryptor.ts:123]: EventType enum?

**Review by t3chguy — COMMENTED** [review]
Looks sane, could we hit 80% covg on the bits outside of rust-crypto at least?

File: src/rust-crypto/RoomEncryptor.ts (hunk 1 above)
Comment (richvdh) [reply src/rust-crypto/RoomEncryptor.ts]: good point, done.

**Comment by richvdh** [comment]
> Looks sane, could we hit 80% covg on the bits outside of rust-crypto at least?

mmkay. I believe we're now there: 14 lines are changing outside rust-crypto, of which 12 are tested.

**Review by t3chguy — APPROVED** [review]

**Pull request merged**